src/Data/Array/Destination.hs

```diff
@@ -179,6 +180,14 @@ fill = Unsafe.toLinear2 unsafeFill
       else
         unsafeDupablePerformIO Prelude.$ MVector.write ds 0 a
 
+-- | @dropEmpty dest@ consumes and empty array and fails otherwise.
+dropEmpty :: HasCallStack => DArray a %1-> ()
+dropEmpty = Unsafe.toLinear unsafeDrop where
+  unsafeDrop :: DArray a -> ()
+  unsafeDrop (DArray ds)
+    | MVector.length ds > 0 = error "Destination.dropEmpty on non-empty array."
+    | otherwise = ()
```

Comment:
You know what? I don't think it's safe. We at least need to `seq` ds like this:
```suggestion
dropEmpty :: HasCallStack => DArray a %1-> ()
dropEmpty = Unsafe.toLinear unsafeDrop where
  unsafeDrop :: DArray a -> ()
  unsafeDrop (DArray ds)
    | MVector.length ds > 0 = error "Destination.dropEmpty on non-empty array."
    | otherwise = ds `seq` ()
```

But generally speaking, I'm wondering if we shouldn't replace `newtype DArray a = DArray (MVector …)` with `data DArray a where { DArray :: MVector … -> DArray a }`. It adds an indirection. But it will save some `Unsafe.toLinear` presumably. Maybe most. This may considerably reduce the kernel of trust, here.

Reply:
I'll make that ^ an issue for later.